src/machine/machine_generic.go

```diff
@@ -48,6 +48,103 @@ func gpioSet(pin Pin, value bool)
 //export __tinygo_gpio_get
 func gpioGet(pin Pin) bool
 
+// Generic PWM/timer peripheral. Properties can be configured depending on the
+// hardware.
+type timerType struct {
+	// Static properties.
+	instance    int32
+	frequency   uint64
+	bits        int
+	prescalers  []int
+	channelPins [][]Pin
+
+	// Configured 'top' value.
+	top uint32
+}
+
+// Configure the PWM/timer peripheral.
+func (t *timerType) Configure(config PWMConfig) error {
+	// Note: for very large period values, this multiplication will overflow.
+	top := config.Period * t.frequency / 1e9
+	if config.Period == 0 {
+		top = 0xffff // default for LEDs
+	}
+
+	// The maximum value that can be stored with the given number of bits in
+	// this timer.
+	maxTop := uint64(1)<<uint64(t.bits) - 1
+
+	// Look for an appropriate prescaler value.
+	var prescaler int
+	for _, div := range t.prescalers {
+		if top/uint64(div) <= maxTop {
```

Comment:
Nit: `if scaled := top/uint64(div); scaled <= maxTop {` and then `top = scaled` below. Not sure the performance and/or the readability gain is worth it, though.